sklearn/neural_network/multilayer_perceptron.py

```diff
@@ -509,7 +537,8 @@ def _fit_stochastic(self, X, y, activations, deltas, coef_grads,
                     activations[0] = X[batch_slice]
                     batch_loss, coef_grads, intercept_grads = self._backprop(
                         X[batch_slice], y[batch_slice], activations, deltas,
-                        coef_grads, intercept_grads)
+                        coef_grads, intercept_grads,
+                        sample_weight[batch_slice])
```

Comment:
`_backprop` is using a subset of the `sample_weights` to weigh the losses. Consider the following sample weights: `[100, 100, 100, 1, 1, 1]` with a batch size is 3, the loss function in `_backprop` will weigh the losses equally.

Reply:
We have considered weighed mini batching, but there was an opinion that this would be okay after enough iterations.

Reply:
If `batch_size=3` and `sample_weight=[100, 100, 100, 1, 1, 1]`, `self._backprop` will be passed a `sample_weight` of `[100, 100, 100]` and `[1, 1, 1]`. When these sample weights are later passed to the loss functions in `self._backprop`, the loss will be weighted uniformly. This will happen in every iteration.

Reply:
So should we adjust the learning rate to account for the overall weight of the mini batch?

Reply:
Hi, if I may, I think the solution would be to accumulate the loss sample wise instead of batch wise.
If so then some more changes in implementation are needed since loss functions are returning scalars, so first we should change them to output numpy arrays. Then `_backprop` should return `batch_loss` also as array, then `_fit_stochastic` would concatenate the `batch_loss` and finally perform weighted average when each epoch ends.

Reply:
@dorcoh Your idea works, but it will change the loss functions API. Since the loss functions are under `_base.py`, I suspect that are private functions, which means we can change their API without a deprecation cycle. Is this correct @jnothman ?